modules/dnn_superres/src/dnn_superres_quality.cpp

```diff
@@ -0,0 +1,470 @@
+
+            Mat img_float;
+            img.convertTo(img_float, CV_32F);
+
+            Mat orig_float;
+            orig.convertTo(orig_float, CV_32F);
+
+            Mat img_diff;
+            absdiff(orig_float, img_float, img_diff);
+
+            Mat img_mul = img_diff.mul(img_diff);
+
+            Scalar sum = cv::sum(img_mul);
+
+            double rgb_sum = 0.0;
+            for(int i = 0; i < 4; i++){
+                rgb_sum += sum[i];
+            }
+
+            double mse = rgb_sum / (double) (3.0 * img_float.total());
+
+            double max = 255 * 255;
+            double psnr = 10 * log10(max / (double) mse);
+
+            return psnr;
+        }
+
+        double DnnSuperResQuality::ssim(Mat img, Mat orig)
+        {
+            CV_Assert(!img.empty());
+            CV_Assert(!orig.empty());
+            CV_Assert(img.type() == CV_8UC3);
+            CV_Assert(img.type() == orig.type());
+
+            Mat ycrcb_img, ycrcb_orig;
+            cvtColor(img, ycrcb_img, COLOR_BGR2YCrCb);
+            cvtColor(orig, ycrcb_orig, COLOR_BGR2YCrCb);
+
+            Mat img_channels[3], orig_channels[3];
+            split(img, img_channels);
+            split(orig, orig_channels);
+
+            Mat img_float;
+            img_channels[0].convertTo(img_float, CV_32F);
+
+            Mat orig_float;
+            orig_channels[0].convertTo(orig_float, CV_32F);
+
+            Mat mu_img, mu_orig;
+            GaussianBlur( img_float, mu_img, Size(11, 11), 0, 0 );
+            GaussianBlur( orig_float, mu_orig, Size(11, 11), 0, 0 );
+            Mat mu_mul = mu_img.mul(mu_orig);
+            mu_img = mu_img.mul(mu_img);
+            mu_orig = mu_orig.mul(mu_orig);
+
+            Mat sigma_img, sigma_orig;
+            GaussianBlur( img_float.mul(img_float), sigma_img, Size(11, 11), 0, 0 );
+            GaussianBlur( orig_float.mul(orig_float), sigma_orig, Size(11, 11), 0, 0 );
+            sigma_img = sigma_img - mu_img;
+            sigma_orig = sigma_orig - mu_orig;
+
+            Mat sigma_mul;
+            GaussianBlur( img_float.mul(orig_float), sigma_mul, Size(11, 11), 0, 0 );
+            sigma_mul = sigma_mul - mu_mul;
+
+            double c1 = (0.01 * 255) * (0.01 * 255);
+            double c2 = (0.03 * 255) * (0.03 * 255);
+
+            Mat nom1 = (2 * mu_mul + c1);
+            Mat nom2 = (2 * sigma_mul + c2);
+
+            Mat nominator = nom1.mul(nom2);
+
+            Mat denom1 = mu_img + mu_orig + c1;
+            Mat denom2 = sigma_img + sigma_orig + c2;
+
+            Mat denominator = denom1.mul(denom2);
+
+            Mat ssim = nominator / denominator;
+
+            Scalar mssim = mean(ssim);
+
+            return mssim[0];
+        }
+
+        void DnnSuperResQuality::benchmark(DnnSuperResImpl sr, Mat img,
+                                            std::vector<double>& psnrValues,
+                                            std::vector<double>& ssimValues,
```

Comment:
I believe this should be moved into a separate sample (module itself is not good place for functions with a lot of customization parameters and/or drawing).

For reproducible benchmark results (in meaning of time) consider using of performance tests which collects multiple samples and drop outliers automatically.

Reply:
I removed it from the module itself, and moved it to just sample codes. I agree that it's better that way. And I added performance test to the module.